service/worker/workerdeployment/workflow.go

```diff
@@ -467,35 +472,68 @@ func (d *WorkflowRunner) handleSetCurrent(ctx workflow.Context, args *deployment
 
 }
 
-func (d *WorkflowRunner) validateAddVersionToWorkerDeployment(version string) error {
+func (d *WorkflowRunner) validateAddVersionToWorkerDeployment(args *deploymentspb.AddVersionUpdateArgs) error {
 	if d.State.Versions == nil {
 		return nil
 	}
 
 	for _, v := range d.State.Versions {
-		if v == version {
+		if v.Version == args.Version {
 			return temporal.NewApplicationError("deployment version already registered", errVersionAlreadyExistsType)
 		}
 	}
 
 	return nil
 }
 
-func (d *WorkflowRunner) handleAddVersionToWorkerDeployment(ctx workflow.Context, version string) error {
+func (d *WorkflowRunner) handleAddVersionToWorkerDeployment(ctx workflow.Context, args *deploymentspb.AddVersionUpdateArgs) error {
 	d.pendingUpdates++
 	defer func() {
 		d.pendingUpdates--
 	}()
-
+	if len(d.State.Versions) >= defaultMaxVersions {
+		err := d.tryDeleteVersion(ctx)
+		if err != nil {
+			return err
+		}
+	}
 	// Add version to local state
 	if d.State.Versions == nil {
-		d.State.Versions = make([]string, 0)
+		d.State.Versions = make([]*deploymentspb.WorkerDeploymentVersionSummary, 0)
 	}
 
-	d.State.Versions = append(d.State.Versions, version)
+	d.State.Versions = append(d.State.Versions, &deploymentspb.WorkerDeploymentVersionSummary{
+		Version:    args.Version,
+		CreateTime: args.CreateTime,
+	})
 	return nil
 }
 
+func (d *WorkflowRunner) tryDeleteVersion(ctx workflow.Context) error {
+	slices.SortFunc(d.State.Versions, func(a, b *deploymentspb.WorkerDeploymentVersionSummary) int {
+		// sorts in ascending order.
+		// cmp(a, b) should return a negative number when a < b, a positive number when a > b,
+		// and zero when a == b or a and b are incomparable in the sense of a strict weak ordering.
+		if a.GetCreateTime().AsTime().After(b.GetCreateTime().AsTime()) {
+			return 1
+		} else if a.GetCreateTime().AsTime().Before(b.GetCreateTime().AsTime()) {
+			return -1
+		}
+		return 0
+	})
+	for _, v := range d.State.Versions {
```

Comment:
Do you want to check for version not being drained here, or that info is not sent to the deployment wf yet?

Reply:
it's not sent to the deployment wf yet, but when I do that I will